### DISCUSSION TITLE
PETSC tests failing on SLURM cluster

### DISCUSSION THREAD

**Discussion by aikubo** [idaholab/moose#20020] [Q&A Getting Started] [answered]
Hello, I am installing MOOSE on our cluster as a cluster user. 

When I try running the PETSC tests using the following lines 

`make PETSC_DIR=/home/akubo/myprojects/moose_models/moose/scripts/../petsc PETSC_ARCH=arch-moose`

I get the following error: 

``` Running check examples to verify correct installation
Using PETSC_DIR=/home/akubo/myprojects/moose_models/moose/petsc and PETSC_ARCH=arch-moose
Possible error running C/C++ src/snes/tutorials/ex19 with 1 MPI process
See http://www.mcs.anl.gov/petsc/documentation/faq.html
[n142][[15633,1],0][btl_openib_component.c:1649:init_one_device] error obtaining device attributes for mlx5_0 errno says Protocol not supported
--------------------------------------------------------------------------
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   n142
  Local device: mlx5_0
--------------------------------------------------------------------------
lid velocity = 0.0016, prandtl # = 1., grashof # = 1.
Number of SNES iterations = 2
Possible error running C/C++ src/snes/tutorials/ex19 with 2 MPI processes
See http://www.mcs.anl.gov/petsc/documentation/faq.html
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 2 slots
that were requested by the application:
  ./ex19

Either request fewer slots for your application, or make more slots available
for use.
--------------------------------------------------------------------------
1,5c1,8
< lid velocity = 0.0016, prandtl # = 1., grashof # = 1.
<   0 SNES Function norm 0.0406612 
<   1 SNES Function norm 4.12227e-06 
<   2 SNES Function norm 6.098e-11 
< Number of SNES iterations = 2
---
> --------------------------------------------------------------------------
> There are not enough slots available in the system to satisfy the 2 slots
> that were requested by the application:
>   ./ex19
> 
> Either request fewer slots for your application, or make more slots available
> for use.
> --------------------------------------------------------------------------
/home/akubo/myprojects/moose_models/moose/petsc/src/snes/tutorials
Possible problem with ex19 running with hypre, diffs above
=========================================
1,9c1,8
< lid velocity = 0.0625, prandtl # = 1., grashof # = 1.
<   0 SNES Function norm 0.239155 
<     0 KSP Residual norm 0.235858 
<     1 KSP Residual norm < 1.e-11
<   1 SNES Function norm 6.81968e-05 
<     0 KSP Residual norm 2.30906e-05 
<     1 KSP Residual norm < 1.e-11
<   2 SNES Function norm < 1.e-11
< Number of SNES iterations = 2
---
> --------------------------------------------------------------------------
> There are not enough slots available in the system to satisfy the 2 slots
> that were requested by the application:
>   ./ex19
> 
> Either request fewer slots for your application, or make more slots available
> for use.
> --------------------------------------------------------------------------
/home/akubo/myprojects/moose_models/moose/petsc/src/snes/tutorials
Possible problem with ex19 running with mumps, diffs above
=========================================
0a1,7
> [n142][[15690,1],0][btl_openib_component.c:1649:init_one_device] error obtaining device attributes for mlx5_0 errno says Protocol not supported
> --------------------------------------------------------------------------
> WARNING: There was an error initializing an OpenFabrics device.
> 
>   Local host:   n142
>   Local device: mlx5_0
> --------------------------------------------------------------------------
/home/akubo/myprojects/moose_models/moose/petsc/src/snes/tutorials
Possible problem with ex19 running with superlu_dist, diffs above
=========================================
Completed test examples
 ```

I have attached the `configure.log` file. 
[configure.log](https://github.com/idaholab/moose/files/7850891/configure.log)
 

I think it might be an issue with SLURM and job allocation? I used the slurm srun command to request and interactive session using one node. But I am not sure how to continue. 

I appreciate any advice or suggestions.

**Comment by GiudGiud** [comment, marked as answer]
Hello

What exactly did you type for srun, and by default how many cores do you get with one node?
It seems to be complaining because it only has one MPI slot and it needs two for the tests. You can change the number of slots (oversuscribe) with the advice given here:
https://stackoverflow.com/questions/35704637/mpirun-not-enough-slots-available

Guillaume

**Reply by aikubo** [reply]
The specific srun command from above error is just:

` srun --account=myaccount --partition=mypartition --x11 --pty --mem=4G bash`

This defaults to one core on one node. 

There are 40 cores per node: 
```
$ grep -c processor /proc/cpuinfo
40
```
Thank you for referring me to that stackexchange I do think that is the issue. However ,the tests still fail. 

I have tried a couple different solutions but still get the same error: 

1. Added the --oversubscribe flag to the /src/snes/tutorial/makefile in the mpi compilation step. This may not be the correct place for this flag so that may be the problem. I have attached the makefile that I edited below. 
[makefile_oversubscribe.txt](https://github.com/idaholab/moose/files/7858936/makefile_oversubscribe.txt)

2. Added the -hostfile flag in the same place. And added localhost slots=1,2,...4 etc
3. Added both --oversubscribe and -hostfile to  /src/snes/tutorial/makefile
4. Invoked a variety of srun environments 
* -n 2 
* -N 1 --cpus-per-task=2
*-n 2 --cpus-per-task=2
5. Tried running it from sbatch using these settings 
	```
	#!/bin/bash
      #SBATCH --nodes=1            ### Number of Nodes
      #SBATCH --ntasks-per-node=2 ### Number of tasks (MPI processes)
      #SBATCH --out=out.out
      #SBATCH --err=err.out
      module load gcc/7.3
      module load openmpi
      
      
      srun ./test_petsc.sh

    ```
* the test_petsc.sh file is just  `make PETSC_DIR=/home/akubo/myprojects/moose_models/moose/scripts/../petsc PETSC_ARCH=arch-moose`

Still getting the same error. I think that I am on the right track right now but maybe I am looking at the wrong makefile. 
Really appreciate your help with this. 

These are the modules loaded: 
```
Currently Loaded Modules:
  1) slurm/19.05   2) cmake/3.12.4   3) gcc/7.3   4) openmpi/2.1

```

**Reply by GiudGiud** [reply]
Hello

I m actually not sure we are on the right track with the oversuscribing now. Finding threads that say this is an OpenFabrics issue
https://users.open-mpi.narkive.com/ha9nLcfD/ompi-error-with-open-mpi-2-0-0-error-obtaining-device-attributes-for-mlx5-0-errno-says-cannot
And this
https://github.com/open-mpi/ompi/issues/8260

Could be tied to openmpi? 2.1 is fairly old. Do you have newer modules available on this cluster?

Guillaume

**Reply by aikubo** [reply]
I found this advice on using the PETSC tests with SLURM and tried it but got the same results 
https://researchcomputing.princeton.edu/support/knowledge-base/petsc

Could be tied to openmpi. I have contacted my cluster admin to see about different versions.

Thank you,  
Allison

**Reply by GiudGiud** [reply]
any update on this?

**Reply by aikubo** [reply]
I was able to get PETSC and libmesh compiled with the following modules 
```
$ module list

Currently Loaded Modules:
  1) slurm/19.05   2) gcc/7.3   3) mpich/3.2   4) python3/3.6.5   5) cmake/3.15.4

```
Thank you for your help!
